src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const stateHistory = [];
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      Object.assign(stateClone, obj.extraData);
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const delObj of obj.keysToRemove) {
+        delete stateClone[delObj];
+      }
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'clear') {
+      for (const key in stateClone) {
+        if (Object.prototype.hasOwnProperty.call(stateClone, key)) {
+          delete stateClone[key];
+        }
+      }
+      stateHistory.push({ ...stateClone });
+    }
+  }
+
+  return stateHistory;
```

Comment:
It is recommended to use a `switch` statement when handling different cases based on the value of `obj.type`. This would make the code more organized and easier to read, and you can also include a `default` case for error handling as a best practice.